### ISSUE TITLE
Todo

### ISSUE THREAD

**Issue by MrScopes** [MrScopes/anilist-js#1]
**THIS TODO IS FOR v1**, TECHNICALLY VALID ON v2

- Better Docs solution
- Better looking README
- Better error handling
- Close to 100% API coverage (so more methods)

**Comment by wiredmatt** [comment]
Aren't you planning on reducing the queries' sizes? They are really big and most data isn't useful honestly, it would be better to have them separated and to fetch specific data, finally having different methods that would: 

- Fetch relevant useful data (id ,name, photo, banner, anime name preferred format)
- Fetch the animes the user is currently watching
- Fetch the animes the user has completed
- Fetch the animes the user is planning to watch
- Fetch the animes the user has dropped
- Fetch the animes the user has paused

If you include an open source license I can do it myself by forking it :)

**Comment by MrScopes** [comment]
> Aren't you planning on reducing the queries' sizes? They are really big and most data isn't useful honestly, it would be better to have them separated and to fetch specific data, finally having different methods that would:
> 
> * Fetch relevant useful data (id ,name, photo, banner, anime name preferred format)
> * Fetch the animes the user is currently watching
> * Fetch the animes the user has completed
> * Fetch the animes the user is planning to watch
> * Fetch the animes the user has dropped
> * Fetch the animes the user has paused
> 
> If you include an open source license I can do it myself by forking it :)

My objective was to keep the api requests limited due to the 90 requests/minute rate limit. What if I did `Media.getStudios()` and it just append the information to `Media.info`. So technically `Media.info.studios` would show up in the interface, but wouldn’t work unless `Media.getStudios().info.studios`, or I could create an InfoManagerStructure for `Media.info.getStudios().nodes` as an example - but this could cause some issues so I definitely want opinions

**Comment by MrScopes** [comment]
another option could be having QueryOptions for each major method,
`.getMedia(1, { size: 'small' })`
`.searchMedia({ search: 'Attack on Titan' }, { size: 'medium' )` with large being the default and containing all info, small containing key info or small amounts of helpful info.

`small:` id, title, description, favourites, coverImage, siteURL

`medium:` small + genres, tags, and studios

`large:` default, and everything possible

**Comment by MrScopes** [comment]
> another option could be having QueryOptions for each major method

the issue is interfaces aren’t adaptive so if you don’t read the tsdoc for the method or the main docs you won’t see that you don’t have the info on small

**Comment by wiredmatt** [comment]
The limit request is per IP using your client id, not for your client id. 

If your library is aiming for server then you should [ask for your client_id to be able to handle more requests](https://anilist.gitbook.io/anilist-apiv2-docs/overview/rate-limiting#rate-limit-raise).

If you are aiming for client-side consumption of the API you will be fine having multiple queries only having the necessary data, there's no way a user makes 90 requests per minute, anyways if it stills get timed out you can handle that.

If you wanted to fetch in-depth data from the studios that made an anime, you would have another query for that, passing in as an argument the id of the media.

I would recommend that the media search query would include relevant data to the user that is doing that: searching, one would expect to get something like what [AniList gives you when you search](https://imgur.com/a/pJtbfQo), and maybe something else like the name of the studio.

Some of your queries end up being a response of more than 500KB.

**Comment by MrScopes** [comment]
> The limit request is per IP using your client id, not for your client id.
> 
> If your library is aiming for server then you should [ask for your client_id to be able to handle more requests](https://anilist.gitbook.io/anilist-apiv2-docs/overview/rate-limiting#rate-limit-raise).
> 
> If you are aiming for client-side consumption of the API you will be fine having multiple queries only having the necessary data, there's no way a user makes 90 requests per minute, anyways if it stills get timed out you can handle that.
> 
> If you wanted to fetch in-depth data from the studios that made an anime, you would have another query for that, passing in as an argument the id of the media.
> 
> I would recommend that the media search query would include relevant data to the user that is doing that: searching, one would expect to get something like what [AniList gives you when you search](https://imgur.com/a/pJtbfQo), and maybe something else like the name of the studio.
> 
> Some of your queries end up being a response of more than 500KB.

I’ll take a further look when I get home in about 2 hours.

**Comment by MrScopes** [comment]
@system32uwu
so yeah to use my idea from earlier it would chain like
```js
console.log(media.getStudios().getAiringSchedule().info.x)
```
or just `media.info.x` wouldn't have studios or airing schedule unless you fetch them

- or maybe use `includeStudios()` etc

**Comment by wiredmatt** [comment]
> or just `media.info.x` wouldn't have studios or airing schedule unless you fetch them

You can make it a minimal query that includes the studio name of each one to show, that can be valuable for the user, about the airing schedule, why would someone need the entire airing schedule? Just get the one for the nextAiringEpisode:

```
{
  Media(id: 113415) {
    id
    coverImage {
      extraLarge
    }
    bannerImage
    status
    averageScore
    description
    startDate{
      year
      month
      day
    }
    title {
      userPreferred
    }
    studios {
      edges {
        node {
          name
        }
      }
    }
    nextAiringEpisode {
      airingAt
      episode
    }
  }
}



```

**Comment by MrScopes** [comment]
> > or just `media.info.x` wouldn't have studios or airing schedule unless you fetch them
> 
> You can make it a minimal query that includes the studio name of each one to show, that can be valuable for the user, about the airing schedule, why would someone need the entire airing schedule? Just get the one for the nextAiringEpisode:
> 
> ```
> {
>   Media(id: 113415) {
>     id
>     coverImage {
>       extraLarge
>     }
>     bannerImage
>     status
>     averageScore
>     description
>     startDate{
>       year
>       month
>       day
>     }
>     title {
>       userPreferred
>     }
>     studios {
>       edges {
>         node {
>           name
>         }
>       }
>     }
>     nextAiringEpisode {
>       airingAt
>       episode
>     }
>   }
> }
> ```

yeah I didn't mean exactly get airing schedule I was just showing you an example, also id would be included with everything including studio

**Comment by MrScopes** [comment]
@system32uwu
check out [v2](https://github.com/MrScopes/anilist.js/tree/v2) - I plan to make a lot of changes there

**Comment by wiredmatt** [comment]
That definitely looks better, keep it up!

**Comment by MrScopes** [comment]
Closed in favor of v3
